c7n/filters/core.py

```diff
@@ -816,3 +823,203 @@ def get_resource_value(self, resource):
         if capture is None:  # regex didn't capture anything
             return None
         return capture.group(1)
+
+
+class ReduceFilter(BaseValueFilter):
+    """Generic reduce filter to group, sort, and limit your resources.
+
+    This example will select the longest running instance from each ASG,
+    then randomly choose 10% of those, maxing at 15 total instances.
+
+    :example:
+
+    .. code-block:: yaml
+
+      - name: oldest-instance-by-asg
+        resource: ec2
+        filters:
+          - "tag:aws:autoscaling:groupName": present
+          - type: reduce
+            group-by: "tag:aws:autoscaling:groupName"
+            sort-by: "LaunchTime"
+            order: asc
+            limit: 1
+
+    Or you might want to randomly select a 10 percent of your resources,
+    but no more than 15.
+
+    :example:
+
+    .. code-block:: yaml
+
+      - name: random-selection
+        resource: ec2
+        filters:
+          - type: reduce
+            order: randomize
+            limit: 15
+            limit-percent: 10
+
+    """
+    annotate = False
+
+    schema = {
+        'type': 'object',
+        # Doesn't mix well with inherits that extend
+        'additionalProperties': False,
+        'required': ['type'],
+        'properties': {
+            # Doesn't mix well as enum with inherits that extend
+            'type': {'enum': ['reduce']},
+            'group-by': {
+                'oneOf': [
+                    {'type': 'string'},
+                    {
+                        'type': 'object',
```

Comment:
if group by is equality, why does it need conversion to compare?

do you have an example use case for regex group by?

Reply:
Group by isn’t just equality. Placing resources into groups is, but when we recombine the groups to get our final list, we sort them the same way things are sorted inside each group. I opted to keep the semantics the same for consistency. 

Someone might want to group and sort resources by the number of ENIs attached or maybe the size of the EBS volume.

As for a regex use case, I could see using part of a value of some tag to group things. It seemed logical to allow for that flexibility.